packages/cloudflare-optimizer-scripts/src/index.js

```diff
@@ -85,7 +87,14 @@ async function handleRequest(event, config) {
     }
   }
 
-  const response = await fetch(request);
+  const response = await fetch(request, { redirect: 'manual' });
+
+
+  // Redirect based statuses should be returned
```

Comment:
@samouri 

That's certainly a question I had for the AMP team

The reason I choose this approach was 

- to stay on the safe side, as this lets the default behavior to persist and client can handle the reaction to redirects based on it's preferences

- We'd need another outgoing fetch request within the worker with redirect mode set to follow to determine the final origin

A downside however, is that it requires extra worker invocation if the final page is still served on the same origin which costs a little extra. 


Do you think we can give the user the ability to chose between the options?

Reply:
> We'd need another outgoing fetch request within the worker with redirect mode set to follow to determine the final origin

Why is that? Because there may be a series of redirects?
Is there an issue with returning the first redirect and allowing the followup redirects to happen in later requests?

> to stay on the safe side, as this lets the default behavior to persist and client can handle the reaction to redirects based on it's preferences

Which client? Are you saying allow for customizing this behavior via the [config](https://github.com/ampproject/amp-toolbox/blob/1abaa6b3f72e517701fb394dd7ef41575c17c2e5/packages/cloudflare-optimizer-scripts/src/index.js#L25-L34) parameter? More customization options is usually a good things, but need to make sure the default makes sense.

----


I'm not sure what the two options are? Here's the situation I'm thinking about:

1. Browser requests `worker.com/example`
2. CF Worker has its `handleEvent` called and attempts to fetch AMP Document from `origin.com/example`
3. Origin returns  a redirect with Location `origin.com/example-new`
4. If CF Worker returns the Response without rewriting the header, the Browser will now redirect to the Origin! This violates expectations. The CF Worker should rewrite the Location header to redirect to `worker.com/example-new` instead.

Reply:
Hi @samouri 

Apologies for the delay in response. 

> I'm not sure what the two options are? Here's the situation I'm thinking about:
> Browser requests worker.com/example
> CF Worker has its handleEvent called and attempts to fetch AMP Document from origin.com/example
> Origin returns a redirect with header Location: origin.com/example-new
> If CF Worker returns the Response without rewriting the header, the Browser will now redirect to the Origin! This violates expectations. The CF Worker should rewrite the Location header to redirect to worker.com/example-new instead.

Thank you very much for providing this example and shedding the light on this case. I was thinking more in terms of the cross-origin redirects where it's not an issue but I get it now. Moreover, this can certainly be even a larger concern in the Proxy mode too, if there's nothing in between the browser.

> Why is that? Because there may be a series of redirects?

> Is there an issue with the most naive approach of returning the first redirect and allowing the followup redirects to happen in later requests?

Since I didn't have the **location rewrite case from origin to worker** in mind, I understood initial question more in terms of intermediary requests being skipped if the final page, happens to be same origin when there are multiple redirects. 

The options I was suggesting were

- Skip the intermediary redirects where possible. 
- Let the Browser itself take care of intermediary redirects

Do you think it's a feasible option or something to be considered, given that it can increase the Egress from the Cloudflare Worker as well as may impact latency and increase complexity of the worker?

Reply:
I will make a PR with Location rewrite in the morning. Have a great week ahead. 👍

Reply:
Just push whenever this is ready and I'll re-review :)

Reply:
Hi @samouri, 

A big apologies for late response, first of all, I will make sure that any future PRs are on time and such delay doesn't occur again. 

---

Now, coming to the PR,

I have made changes, with regards to 

- Rewriting Location Header for Redirects (both GET and non-GET) 
- The above rewriting logic is done only in Proxy mode 

Since in the Cloudlfare Routes mode, Cloudflare itself is intercepting the requests, the redirects should be intercepted as well, so I only implemented this functionality for Proxy mode. 

Can you think of any valid cases where Routes mode might require Location rewriting as well?

I can implement it, if there are any cases that you can think of. In that case, should we read the hostname from the request body, or require the user to give an origin in configuration regardless of whether it's Proxy or Route mode? 


---- 

Thank you so much for reviewing the PR and spending your valuable time reviewing the PR. :heart: 
 
--- 

Looking forward to make more PRs to this repository in future without the horrendous delays